src/internal/document_api/mod.rs

```diff
@@ -685,6 +738,95 @@ fn process_users(
     (user_errs, users_with_key)
 }
 
+fn process_policy(
+    policy_result: &PolicyResult,
+) -> (Vec<DocAccessEditErr>, Vec<WithKey<UserOrGroup>>) {
+    let (de_errs, policy_eval_results): (Vec<DocAccessEditErr>, Vec<WithKey<UserOrGroup>>) =
+        policy_result
+            .user_or_groups()
+            .into_iter()
+            .partition_map(|uog| match uog {
+                UserOrGroupWithKey::User {
+                    id,
+                    master_public_key: Some(key),
+                } => {
+                    let user = UserOrGroup::User {
+                        // okay since these came back from the service
+                        id: UserId::unsafe_from_string(id.clone()),
+                    };
+
+                    Either::from(
+                        key.clone()
+                            .try_into()
+                            .map(|k| WithKey::new(user.clone(), k))
+                            .map_err(|_e| {
+                                DocAccessEditErr::new(
+                                    user,
+                                    format!("Error parsing user public key {:?}", &key),
+                                )
+                            }),
+                    )
+                }
+                UserOrGroupWithKey::Group {
+                    id,
+                    master_public_key: Some(key),
+                } => {
+                    let group = UserOrGroup::Group {
+                        // okay since these came back from the service
+                        id: GroupId::unsafe_from_string(id.clone()),
+                    };
+
+                    Either::from(
+                        key.clone()
+                            .try_into()
+                            .map(|k| WithKey::new(group.clone(), k))
+                            .map_err(|_e| {
+                                DocAccessEditErr::new(
+                                    group,
+                                    format!("Error parsing group public key {:?}", &key),
+                                )
+                            }),
+                    )
+                }
+
+                any => Either::Left(DocAccessEditErr::new(
+                    any.clone().into(),
+                    "User or group does not have associated public key".to_string(),
```

Comment:
I know it's a bit more verbose, but the caller won't know which one didn't have a key. Is it a user or is it a group?

Reply:
True. We have a nice Display for UserOrGroup, so I reworked it to use that.